### DISCUSSION TITLE
interface support?

### DISCUSSION THREAD

**Discussion by ZeeD** [rokucommunity/brighterscript#671] [General]
Poking in the source internals, I found some references on a way to set up "interfaces"; for example there are [some tests](https://github.com/rokucommunity/brighterscript/blob/master/src/parser/tests/statement/InterfaceStatement.spec.ts ) with a syntax like

```bs
interface Person
    public as string
    protected as string
    private as string
    sub as string
    function as string
    interface as string
    endInterface as string
end interface
```

I don't see any references on `interface` on the docs, and if I try to write one, I see that `bsc` doesn't give me errors, but it seems that the interface definition "disappears" in the generated `.brs` files.

Is this a supported feature from the language? will it become one? (did I misunderstood it?)

**Comment by TwitchBronBron** [comment]
I keep meaning to write up some docs about interfaces...sorry about that.

So, interfaces are officially supported in brighterscript right now. They're a compile-time only construct, so they don't exist at runtime (which is why they disappear from the transpiled code). 

Right now, you can use them as class property types, interface property types, and function parameter types. 

![image](https://user-images.githubusercontent.com/2544493/185950491-c6ff0b8c-262b-4b3d-9c45-16cb7ed74a34.png)

**Reply by ZeeD** [reply]
Oh, ok!
so they are some kind of [typescript interfaces](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#interfaces), and not a way to write [roku interfaces](https://developer.roku.com/de-de/docs/references/scenegraph/xml-elements/interface.md)!
Cool, I can use them :)

**Reply by TwitchBronBron** [reply]
Correct! We intend to use them much more extensively in the future. Like adding interfaces that describe all `createObject()` object types, components, etc. And enhancing third party library information to include interfaces in type definitions, as well as validations in all the interface usages across your app. Just not enough time in the day. :D
